docs/rfcs/xxxx-integrations-in-operator.md

```diff
@@ -0,0 +1,234 @@
+---
+
+# Integrations in Grafana Agent Operator
+
+* Date: 2022-01-04
+* Author: Robert Fratto (@rfratto)
+
+## Background
+
+At the time of writing, Grafana Agent Operator can be used to collect metrics
+and logs from a Kubernetes cluster by deploying and managing Grafana Agent
+pods. However, logs and metrics are just a subset of the total functionality
+available in Grafana Agent. This document proposes adding support for
+integrations into the Grafana Agent Operator.
+
+This proposal supersedes [#883][], which was the first attempt at designing the
+feature. This proposal takes advantage of the lessons we've learned and
+utilizes the upcoming [integrations revamp][#1198] work done in December 2021
+so this functionality can be implemented more easily.
+
+## Goals
+
+* Allow Grafana Agent Operator to deploy integrations
+* Allow Grafana Agent Operator to collect metrics from deployed integrations
+* Support both per-node and "normal" integrations
+* Use a generic CRD for all metrics-based integrations
+
+The final goal prevents us from incurring any additional overhead when
+developing new integrations. The operator will be able to configure and deploy
+any integration without having to know what the underlying Grafana Agent
+version supports.
+
+## Non-Goals
+
+* Minimize number of created pods used for metrics / logging / integrations
+* Discussion of deployment strategies for multiple GrafanaAgents
+
+## Proposal
+
+At a high level, the proposal is to:
+
+* Define two new CRDs:
+  * `MetricsIntegration`: an integration to run
+  * `IntegrationMonitor`: collection rules for integrations, similar to
+    ServiceMonitor.
+* Update `GrafanaAgent` to discover `MetricsIntegration`s and run integrations.
+* Update `MetricsInstance` to discover `IntegrationMonitor`s and collect
+  metrics from integrations.
+
+## Architecture
+
+### Running integrations
+
+The new CRD, MetricsIntegration, will be used for supporting all metrics-based
+integrations today. A metrics-based integration is an integration that exposes
+metrics. These integrations have an always-available set of common
+configuration fields. We define MetricsIntegration as a CRD specific for
+metrics-based integrations to add guarantees that these common configuration
+fields can be used when necessary for implementation details.
+
+The spec of MetricsIntegration primarily revolves around three fields:
+
+* `name`: The name of the integration (e.g., `node_exporter`, `mysqld_exporter`)
+* `type`: The type of integration
+* `config`: A YAML string declaring the configuration for the integration.
+
+`type` is critical for determining how the integration should be deployed.
+There are three supported values for `type`:
+
+* `daemonset`: Declares that the `name` integration should be run on every Node in the
+  Kubernetes cluster. It is invalid to have a GrafanaAgent CRD discover more
+  than one `daemonset` integrations with the same `name` . Example integration:
+  `node_exporter`.
```

Comment:
While practical for sure, I'm not convinced we should ship all those exporters in the grafana-agent binary at all. It's probably perfectly okay to tell people to install `prometheus-node-exporter` from https://prometheus-community.github.io/helm-charts with the values necessary to get monitors installed and labels in the right shape dashboards expect them to be (and document this prominently)

Reply:
We'll probably always want to continue embedding _some_ exporters in the agent, since it makes things easier for non-Kubernetes users, which is still a userbase we're interested in helping. 

Though if an exporter only makes sense for Kubernetes (or isn't popular enough to justify embedding it), we would definitely prefer recommending installing that exporter with Helm and using a ServiceMonitor/PodMonitor/Probe for collecting metrics from it instead of embedding it directly into the agent.